draft-ietf-core-problem-details.md

```diff
@@ -13,224 +17,297 @@ stand_alone: yes
+This document defines a "problem detail" as a way to carry
+machine-readable details of errors in a REST response to avoid the
+need to define new error response formats for REST APIs.
+The format
+is inspired by, but intended to be more concise than, the Problem
+Details for HTTP APIs defined in RFC 7807.
 
 --- middle
 
 # Introduction
 
-CoAP {{!RFC7252}} response codes are sometimes not sufficient to convey enough information about an error to be helpful.  This specification defines a simple and extensible CoRAL {{!I-D.ietf-core-coral}} vocabulary to suit this purpose.  It is designed to be reused by CoAP APIs, which can identify distinct "problem types" specific to their needs.  Thus, API clients can be informed of both the high-level error class (using the response code) and the finer-grained details of the problem (using this vocabulary), as shown in {{fig-problem-details}}.
-
-~~~
-{::include problem-details.ascii-art}
+REST response status information such as CoAP {{-coap}} response
+codes is sometimes not sufficient to convey enough information about
+an error to be helpful.  This specification defines a simple and
+extensible framework to define CBOR tags to suit this purpose.
+It is designed to be reused by REST APIs, which can identify distinct
+"problem types" specific to their needs.
+Thus, API clients can be informed of both the high-level error class
+(using the response code) and the finer-grained details of the problem
+(using this vocabulary), as shown in {{fig-problem-details}}.
+
+~~~ aasvg
++--------+           +--------+
+|  CoAP  |           |  CoAP  |
+| Client |           | Server |
++----+---+           +----+---+
+     |                    |
+     | Request            |
+     |------------------> |
+     |                    |
+     | <----------------- |
+     | Error Response     |
+     | with a CBOR Data   |
+     | Item giving        |
+     | Problem Details    |
+     |                    |
 ~~~
-{: #fig-problem-details artwork-align="center" title="Problem Details"}
+{: #fig-problem-details artwork-align="center"
+   title="Problem Details: Example with CoAP"}
 
-The vocabulary presented is largely inspired by the Problem Details for HTTP APIs defined in {{?RFC7807}}.
+The framework presented is largely inspired by the Problem Details for HTTP APIs defined in {{RFC7807}}.
 
 ## Requirements Language
 
-{::boilerplate bcp14}
+{::boilerplate bcp14-tagged}
 
 # Basic Problem Details
 
-A CoRE Problem Details is a CoRAL document with the following elements:
+A Concise Problem Details data item is a CBOR data item with the following
+structure (notated in CDDL {{!RFC8610}}, using 65535 in place of a tag number to
+be defined for the type of problem details):
+
+~~~ CDDL
+problem-details = #6.65535(problem-details-map)
+problem-details-map = {
+  ? &(title: -1) => text
```

Comment:
Hi Thomas,

in the current draft, `title` is not a URI, but human-readable text, so I don't understand several of your comments.  In the PR, there actually is no way to put in a URI for the problem type (≠ title).

Reply:
Arg, sorry, I need more powerful reading glasses apparently.  I merged "type" and "title" into one thing.  Please re-interpret all my comments as "can we rather use an explicit URI `type` attribute instead of the CBOR tag to handle extensibility?"

Reply:
Of course we could.  The idea with the PR was that many of these types work better if registered, and we have a nice registry that we can use for that.

Reply:
UX question: are we saying that *all* extensions (including `custom-problem-detail-entries`) need to go through IANA, or does that apply to `standard-problem-detail-entries` only?

Reply:
On standard problem details (which is what makes them different from custom ones), not custom ones (these are defined by the type).
We should say this more explicitly.

Reply:
> We should say this more explicitly.

In sec-new-attributes we say _"To establish a *custom* problem type, A CBOR Tag number needs to be registered"_, which is confusing given we are referring to `standard-problem-detail-entries` rather than `custom-problem-detail-entries`

Reply:
Regarding handling `custom-problem-detail-entries`: it is not clear to me how a receiver is supposed to distinguish between custom map entry "A" (e.g., w/ key 10) from custom map entry "B" (again w/ key 10) without an explicit type/sub-type/namespacing indication.

Reply:
> In sec-new-attributes we say "To establish a custom problem type, A CBOR Tag number needs to be registered", which is confusing given we are referring to standard-problem-detail-entries rather than custom-problem-detail-entries

I think "custom problem type" was intended to be in contrast with the predefined Generic Problem Details type. The sentence can be rephrased as "To establish a new problem type different than the Generic Problem Details type defined in this specification, a CBOR Tag number needs to be registered".

Reply:
> Regarding handling custom-problem-detail-entries: it is not clear to me how a receiver is supposed to distinguish between custom map entry "A" (e.g., w/ key 10) from custom map entry "B" (again w/ key 10) without an explicit type/sub-type/namespacing indication.

As I understand, the new tag comes with exactly one corresponding unique uint/tstr key, so there is no risk of collision as long as uniqueness is ensured (bottom line, when registering the new tag). As defined in sec-new-attributes:

"Problem type definitions MAY extend the Problem Details document with additional entries to convey additional, problem-specific information. These receive a custom problem detail entry map key (unsigned integer
or text) and SHOULD be described in a specification that goes along with such a registration.  Such a specification SHOULD also reference the present specification."

About the uniqueness of those map keys, this would mean that the review of a new tag registration has to also check that the new specification is indeed introducing a still available entry map key for the new custom problem detail. Is there a better way than checking all the documents that have registered a tag for custom problem details? Perhaps the map key can be mentioned in the "Semantics" column of the tag registry when registering the tag?